### PR TITLE
z80scc: basic wait/dma request support

### PR DESCRIPTION
**Pull request by pmackinlay**
I've done the absolute minimum necessary to get serial DMA working for the InterPro.

This involved implementing some unimplemented logic around the SCC "wait/dma request" behaviour on transmit. I have not attempted to tackle either the "wait function" or receive, so this is partial at best. It's also possible that there are some other points where the transmit buffer can be filled/emptied and the new check_waitrequest() function needs to be called.

Appreciate a review from those with more detailed understanding of the SCC device, but it's working for me at the moment and I think there are few side-effects.

### CODE REVIEW

File: src/devices/machine/z80scc.h [hunk 1]
```diff
@@ -631,6 +631,8 @@ class z80scc_channel : public device_t,
 	// SCC specifics
 	int m_ph;       // Point high command to access regs 08-0f
 	uint8_t m_zc;
+
+	void check_waitrequest();
```
Comment (cuavas) [inline src/devices/machine/z80scc.h]: Can you please move this up to be with the other private member function declarations?  Intermixing member functions and data members can get confusing.

File: src/devices/machine/z80scc.cpp [hunk 2]
```diff
@@ -2857,3 +2863,32 @@ WRITE_LINE_MEMBER(z80scc_channel::write_rx)
...
+	if (m_wr1 & WR1_WRDY_ON_RX_TX)
+		return;
+
+	// if dma request function is enabled
+	if (m_wr1 & WR1_WRDY_FUNCTION)
+	{
+		// assert /W//REQ if transmit buffer is empty, clear if it's not
+		int state = (m_rr0 & RR0_TX_BUFFER_EMPTY) ? ASSERT_LINE : CLEAR_LINE;
+
+		m_index ? m_uart->m_out_wrdyb_cb(state) : m_uart->m_out_wrdya_cb(state);
```
Comment (cuavas) [inline src/devices/machine/z80scc.cpp]: These are both write line devcb, aren't they?  Remember C++ `?:` operator can yield an lvalue - you can probably simplify this line slightly to:
```
(m_index ? m_uart->m_out_wrdyb_cb : m_uart->m_out_wrdya_cb)(state);
```
It's not much simpler, but it only says `state` once.

**Comment by pmackinlay** [comment]
Made changes requested by @cuavas. Also renamed WRDY to WREQ to match Zilog documentation.

@JoakimLarsson indicated this didn't seem too crazy, but that he would like to rework in future. There is a significant likelihood this patch doesn't follow the precise DMA mode timing described in the Zilog documentation, however as I mentioned before, as a preliminary implementation, it's good enough for my own immediate need, and I hope it can be committed on that basis.

**Comment by JoakimLarsson** [comment]
Yes I have DMA support for SCC on the TODO list so I will go over the data sheet regarding timings etc, I'd just hoped to find a more thorough test case then 1 byte transfers :) I'll also align the coding style etc when I am at it. Thanks for the contribution!

**Comment by JoakimLarsson** [comment]
What is the startup sequence to trigger the DMA line to set de-/asserted? I have put some debug statements in there that doesn't seem to be used, guess I need a disk image and some proper command line to get it going?

**Comment by pmackinlay** [comment]
Yes, you'll need to get the ip2800 machine running with the "boot floppy" image in the 3.5" floppy drive. You can get an image from here: http://web.archive.org/web/20041108210810/http://starfish.osfn.org:80/Intergraph/floppies/2500/ (the boot.1 file), but you'll have to expand it to exactly 1474560 bytes - you can use "truncate -s" under msys2 to do it.

After that, it should be possible to boot to FDM, and then enter "8 0 1 boot" and wait a long time. A single message similar to "booting system" will be output via serial dma. The rest of the boot sequence hangs, so don't expect more.

**Pull request merged**